src/tools/coverage-dump/src/parser.rs

```diff
@@ -0,0 +1,116 @@
+#[cfg(test)]
+mod tests;
+
+use anyhow::{anyhow, ensure};
+use regex::bytes;
+use std::sync::OnceLock;
+
+/// Given the raw contents of a string literal in LLVM IR assembly, decodes any
+/// backslash escapes and returns a vector containing the resulting byte string.
+pub(crate) fn unescape_llvm_string_contents(contents: &str) -> Vec<u8> {
+    let escape_re = {
+        static RE: OnceLock<bytes::Regex> = OnceLock::new();
+        // LLVM IR supports two string escapes: `\\` and `\xx`.
+        RE.get_or_init(|| bytes::Regex::new(r"\\\\|\\([0-9A-Za-z]{2})").unwrap())
+    };
+
+    fn u8_from_hex_digits(digits: &[u8]) -> u8 {
+        // We know that the input contains exactly 2 hex digits, so these calls
+        // should never fail.
+        assert_eq!(digits.len(), 2);
+        let digits = std::str::from_utf8(digits).unwrap();
+        u8::from_str_radix(digits, 16).unwrap()
+    }
+
+    escape_re
+        .replace_all(contents.as_bytes(), |captures: &bytes::Captures<'_>| {
+            let byte = match captures.get(1) {
+                None => b'\\',
+                Some(hex_digits) => u8_from_hex_digits(hex_digits.as_bytes()),
+            };
+            [byte]
+        })
+        .into_owned()
+}
+
+pub(crate) struct Parser<'a> {
+    rest: &'a [u8],
+}
+
+impl<'a> Parser<'a> {
+    pub(crate) fn new(input: &'a [u8]) -> Self {
+        Self { rest: input }
+    }
+
+    pub(crate) fn ensure_empty(self) -> anyhow::Result<()> {
+        ensure!(self.rest.is_empty(), "unparsed bytes: 0x{:02x?}", self.rest);
+        Ok(())
+    }
+
+    pub(crate) fn read_n_bytes(&mut self, n: usize) -> anyhow::Result<&'a [u8]> {
+        ensure!(n <= self.rest.len());
+
+        let (bytes, rest) = self.rest.split_at(n);
+        self.rest = rest;
+        Ok(bytes)
+    }
+
+    pub(crate) fn read_uleb128_u32(&mut self) -> anyhow::Result<u32> {
+        let (_bytes, value, rest) = split_uleb128_u64(self.rest)?;
+        let value: u32 = value.try_into()?;
+        self.rest = rest;
+        Ok(value)
+    }
+
+    pub(crate) fn read_uleb128_usize(&mut self) -> anyhow::Result<usize> {
+        let (_bytes, value, rest) = split_uleb128_u64(self.rest)?;
+        let value: usize = value.try_into()?;
+        self.rest = rest;
+        Ok(value)
+    }
+}
+
+/// Splits a byte slice into two parts: a prefix containing a ULEB128-encoded
+/// `u64` value, and the rest of the slice. The decoded value is also returned.
+///
+/// Returns `None` if the input doesn't start with a complete ULEB128 value,
+/// or if the value is too large to fit in a `u64`.
+fn split_uleb128_u64(input: &[u8]) -> anyhow::Result<(&[u8], u64, &[u8])> {
```

Comment:
Surprisingly, the compiler currently doesn’t use `leb128`. It does use `gimli`, but it turns out that gimli uses its own internal copy of leb128 rather than the published crate. But they both seem to be published by the same org, so I imagine it wouldn’t be controversial to add the dependency.

Reply:
In earlier prototypes I was printing out the raw bytes associated with each line’s values, which is why I included the first slice in the return value.

That functionality didn’t end up in the current PR (since it was a lot of complexity and made the dumps noisier), but by that point I didn’t see any strong reason to change the underlying uleb code to get rid of the unused slice.